### ISSUE TITLE
aws-cdk: stack difference count on diff is incorrect

### ISSUE THREAD

**Issue by jchli** [aws/aws-cdk#27238]
### Describe the bug

The reported number of stack differences on `cdk diff` is incorrect

### Expected Behavior

A stack with no diffs (e.g. just deployed cleanly) should report 0 differences

### Current Behavior

```
% AWS_PROFILE=non-prod npm run cdk diff my-app -- --exclusively --context myenv=stage --fail

> aws-cdk-infra@0.1.0 cdk
> cdk diff my-app --exclusively --context myenv=stage --fail

[WARNING] aws-cdk-lib.aws_rds.DatabaseClusterProps#instanceProps is deprecated.
  - use writer and readers instead
  This API will be removed in the next major release.
Stack my-app (my-app-stage)
There were no differences

✨  Number of stacks with differences: 2
```

### Reproduction Steps

We're running aws-cdk and aws-cdk-lib on 2.96.2

Our `my-app` stack includes **nested stacks** such as aurora postgres, ECR, s3, ECS fargate. I can provide more detailed/redacted CDK code if you aren't able to reproduce

### Possible Solution

It seems like this might be related to this: https://github.com/aws/aws-cdk/pull/26796, which was seemingly fixed in 2.93.0.

More precisely, likely something to do with this line which adds up stacks and nested stacks: https://github.com/aws/aws-cdk/pull/26796/files#diff-8cff125f4b53b025add8740ab74f39ee6c20221c0d3ac5ab7db184a6bb077398R154

### Additional Information/Context

I just ran into this bug on 2.96.2 but also tested 2.93.0 with the same results.

### CDK CLI Version

2.96.2

### Node.js Version

18.17.1

### OS

Darwin jasonliu macbook 22.4.0 Darwin Kernel Version 22.4.0: Mon Mar  6 21:00:17 PST 2023; root:xnu-8796.101.5~3/RELEASE_X86_64 x86_64

### Language

Typescript

### Language Version

typescript 5.1.6

**Comment by indrora** [comment]
Thank you for the report.

**Comment by ChaseWagoner** [comment]
@indrora I generated a sample app with `npx cdk init sample-app --language=typescript`, then modified it to reproduce the issue:

```typescript
// cdk-test-stack.ts
import { NestedStack, Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';

export class CdkTestStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);
    new TestNestedStack(this, 'CdkTestNestedStack');
  }
}

class TestNestedStack extends NestedStack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);
  }
}
```

Testing:
```sh
$ cdk deploy CdkTestStack # Succeeds, creating both stacks

$ cdk diff --fail
Stack CdkTestStack
There were no differences

✨  Number of stacks with differences: 1

$ echo $?
1
```

**Comment by jortkoopmans** [comment]
I have the same issue, which can be reproduced with the steps described by @ChaseWagoner. One minor fix to the reproduction steps: you need to modify the import at the `bin/cdk-testapp.ts` to point to correctly point to CdkTestStack (instead of default CdkTestappStack). Or you could alternatively change the class name to match. Then it should synth/deploy OK.

I'm (also) quite certain it is directly related to issue #26818 with PR #26796 , as I have successfully tested with version `2.92.0` where this issue does not occur. As soon as you flip to `2.93.0` or higher (up until `2.104.0`), it will be broken.

**Comment by jchli** [comment]
hey @SankyRed - any movement on this?

**Comment by SankyRed** [comment]
After testing the issue locally seems like the issue has been fixed. The below image shows when a nested stack has a change which shows the diff as 1.
<img width="1060" alt="Screenshot 2024-01-08 at 2 45 58 PM" src="https://github.com/aws/aws-cdk/assets/121981451/3825d6ba-4eff-473b-8705-7ca92f82b4d0">
Whereas after deploying the stack, when we run `cdk diff` it shows the diff count as 0 and works as expected. 
<img width="967" alt="Screenshot 2024-01-08 at 2 46 06 PM" src="https://github.com/aws/aws-cdk/assets/121981451/48a08644-aba4-4e1f-88b9-f9ed54b6305f">
Closing the issue. Thank you!

**Comment by ChaseWagoner** [comment]
@SankyRed Out of curiosity, which version of the CDK packages did you use for testing? The [minimal test](https://github.com/aws/aws-cdk/issues/27238#issuecomment-1737993737) from before still shows an incorrect diff stack count. (I also reproduced the incorrect diff count with a realistic stack containing various resources and nested stacks.)

I still see the suspect line in `main` branch, which adds the nested stack count as "diffs," seemingly without regard for whether those stacks actually contain a diff: https://github.com/aws/aws-cdk/blob/8562c173227f7b1d522ffaca92338f54d4bb1a01/packages/aws-cdk/lib/cdk-toolkit.ts#L180

#### Setup

- `npx cdk init sample-app --language=typescript`
- Modify `cdk-test-stack.ts` to match my previous comment

#### Test 2.92.0

```sh
npm i aws-cdk@2.92.0
npm i -D aws-cdk-lib@2.92.0
npm run cdk synth
npm run cdk deploy
npm run cdk diff -- --fail
echo $?
```

Output shows no diff, as expected:
```sh
> cdk-test@0.1.0 cdk
> cdk diff --fail

Stack CdkTestStack
There were no differences

✨  Number of stacks with differences: 0

0 # exit code
```

#### Test 2.118.0

```sh
npm i aws-cdk@latest
npm i -D aws-cdk-lib@latest
npm ls aws-cdk aws-cdk-lib
# cdk-test@0.1.0
# ├── aws-cdk-lib@2.118.0
# └── aws-cdk@2.118.0
npm run cdk synth
npm run cdk deploy
npm run cdk diff -- --fail
echo $?
```

Output shows an incorrect diff, including every nested stack in the diff count (even those without diffs):
```sh
> cdk-test@0.1.0 cdk
> cdk diff --fail

Stack CdkTestStack
There were no differences

✨  Number of stacks with differences: 1

1 # exit code
```

**Comment by SankyRed** [comment]
@ChaseWagoner Apologies on that, tested the dependent stack instead of nested stack earlier. When run `cdk diff` with stacks containing nested stacks with no changes the issue seems to persist. Thanks for calling it out.
<img width="769" alt="Screenshot 2024-01-08 at 4 09 33 PM" src="https://github.com/aws/aws-cdk/assets/121981451/23b62d1b-0699-428e-984e-2e6d85505579">

**Comment by Rondineli** [comment]
I still facing this issue, did anybody manage to fix/bypass it? Running on latest version (2.127.0).

**Comment by jchli** [comment]
Hey @SankyRed - any news on this? This blocks us from upgrading to a newer version since we use the diff count as part of our release process.
